react-carousel/my-app/src/Carousel.tsx

```diff
@@ -0,0 +1,44 @@
+import { useEffect, useState } from 'react';
+import { Button } from './Button';
+import { Image } from './Image';
+
+type Prop = {
+  imgRefs: string[];
+};
+
+export function Carousel({ imgRefs }: Prop) {
+  const [state, setState] = useState(0);
+  const [timer, setTimer] = useState<NodeJS.Timeout>();
+  useEffect(() => {
+    setTimer(
+      setTimeout(() => {
+        setState((state + 1) % imgRefs.length);
+      }, 3000)
+    );
+    return clearTimeout(timer);
```

Comment:
This doesn't work like you expect. You are actually clearing the timer right here every time the effect runs. You should return a _function_ as your clean-up. Since you are not returning a function, your clean-up is not being called by React. However, since you are calling the function here in line 18, it is being executed every time the effect runs.

You are somewhat lucky here, though, because you are calling `clearTimeout(timer)` and passing it the _previous_ value of `timer` (that is, the one before you call `setTimer` on line 13), because `setTimer` is just updating the cache and not modifying the `timer` variable. So you are actually clearing the previous timer, which is what you want.

So, while the code is wrong, it may by a fortunate accident be doing what you want.

Reply:
I'll have to be very careful about this in the future. I knew I would be referring to the previous timer, that's why I saved it with the useState hook, but I didn't catch how narrowly I avoided using the variable before I updated it, thanks for pointing that out!